File: scray-querying/modules/scray-sync-file/src/main/scala/scray/sync/impl/FileVersionedDataApiImpl.scala
// See the LICENCE.txt file distributed with this work for additional
// information regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// See the LICENCE.txt file distributed with this work for additional
// information regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package scray.sync.impl

import scray.sync.api.{VersionedData, VersionedDataApi}

import scala.collection.mutable.HashMap
import java.util.ArrayList
import com.google.gson.Gson
import com.google.gson.GsonBuilder

import java.io.FileWriter
import java.io.BufferedWriter
import java.io.File
import com.google.gson.reflect.TypeToken
import scray.sync.api.VersionedData

import collection.JavaConverters._
import com.typesafe.scalalogging.LazyLogging

import java.io.OutputStream
import java.io.OutputStreamWriter
import java.io.InputStream
import java.util
import scala.io.Source

class FileVersionedDataApiImpl extends VersionedDataApi with LazyLogging {
  var versionInformations: HashMap[Int, VersionedData] = this.toMap(new ArrayList[VersionedData])
  val gson = new GsonBuilder().setPrettyPrinting().create();

  def getLatestVersion(dataSource: String, mergeKey: String): Option[VersionedData] = {
    versionInformations.get(scray.sync.api.VersionedData.createVersionKey(dataSource, mergeKey))
  }

  def updateVersion(dataSource: String, mergeKey: String, version: Long, data: String) {
    versionInformations.put(VersionedData.createVersionKey(dataSource, mergeKey), new VersionedData(dataSource, mergeKey, version, data))
  }

  def updateVersion(vd: VersionedData) {
    versionInformations.put(VersionedData.createVersionKey(vd.dataSource, vd.mergeKey), vd)
  }

  def persist(path: String): Unit = {
    this.writeToFile(path)
  }
  
  def persist(outStream: OutputStream): Unit = {
    this.writeToFile(outStream)
  }
  
  def load(path: String): Unit = {
    versionInformations = readFromFile(path)
  }
  
  def load(stream: InputStream): Unit = {
    versionInformations = readFromInputStram(stream)
  }

  private def readFromFile(path: String): HashMap[Int, VersionedData] = {

    val dataList = () => {
      try {
        val fileContents = Source.fromFile(path).getLines.mkString

        val jsonParser = new GsonBuilder().create();
        val listType = new TypeToken[ArrayList[VersionedData]]() {}.getType();

        gson.fromJson(fileContents, listType)
      } catch {
        case _ => {
          logger.debug(s"Unable to open file ${path}. Use empty verson collection")
          new ArrayList[VersionedData]
        }
      }
    }
    toMap(dataList())
  }
  
    private def readFromInputStram(stream: InputStream): HashMap[Int, VersionedData] = {

    val dataList = () => {
      try {
        val fileContents = Source.fromInputStream(stream).getLines.mkString

        val jsonParser = new GsonBuilder().create();
        val listType = new TypeToken[ArrayList[VersionedData]]() {}.getType();

        gson.fromJson(fileContents, listType)
      } catch {
        case _ => {
          logger.debug(s"Unable to read from file. Use empty verson collection")
          new ArrayList[VersionedData]
        }
      }
    }
    toMap(dataList())
  }

  private def toMap(dataList: java.util.List[VersionedData]) = {
    dataList.asScala.foldLeft(new HashMap[Int, VersionedData])((acc, nexVersiondata) => {
      val keyExists = acc.get(nexVersiondata.getVersionKey) match {
        // Add data if not exists
        case None => acc.put(nexVersiondata.getVersionKey, nexVersiondata)
        case Some(existingVersion) => {
          if (existingVersion.version > nexVersiondata.version) {
            logger.debug(s"Existing version is newer than next version. Existing: ${existingVersion}, Next: ${nexVersiondata}")
          } else {
            acc.put(nexVersiondata.getVersionKey, nexVersiondata)
          }
        }
      }
      acc
    })
  }

  private def toList(versionInformations: HashMap[Int, VersionedData]): java.util.List[VersionedData] = {
    versionInformations.keySet.map(key => {
      versionInformations.get(key).get
    }).toList.asJava
  }

  private def writeToFile(path: String): Unit = {
    val jsonString = gson.toJson(toList(versionInformations))

    val file = new File(path)
    val bw = new BufferedWriter(new FileWriter(file))
    bw.write(jsonString)
    bw.close()
  }
  
  private def writeToFile(outStream: OutputStream): Unit = {
    val jsonString = gson.toJson(toList(versionInformations))

    val bw = new OutputStreamWriter(outStream)
    bw.write(jsonString)
    bw.flush()
    bw.close()
  }

  /**
   * Get all resources where a version exits for
   */
  override def getAllVersionedResources(): util.List[VersionedData] = {
    this.toList(this.versionInformations);
  }
}

File: scray-querying/modules/scray-sync-file/src/main/java/scray/sync/impl/FileVersionedDataApiImpl.java
package scray.sync.impl;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import scala.Array;
import scray.sync.api.VersionedData;
import scray.sync.api.VersionedDataApi;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.Map.Entry;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

public class FileVersionedDataApiImpl implements VersionedDataApi {
    private static final Logger logger = LoggerFactory.getLogger(FileVersionedDataApiImpl.class);

    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private Map<Integer, VersionedData> versionInformations = new ConcurrentHashMap<>();
    private Map<Integer, Map<Integer, VersionedData>> versionInformationsIdxEnvState = new ConcurrentHashMap<>();
    private Idx idx = new IndexDataEnvState();

    private IndexDataEnvState indexCreator = new IndexDataEnvState();


    public FileVersionedDataApiImpl() {
        this.versionInformations = toMap(new ArrayList<>());
    }

    @Override
    public Optional<VersionedData> getLatestVersion(String dataSource, String mergeKey) {
        int key = VersionedData.createVersionKey(dataSource, mergeKey);
        return Optional.ofNullable(versionInformations.get(key));
    }

    @Override
    public void updateVersion(String dataSource, String mergeKey, long version, String data) {
        VersionedData vd = new VersionedData(dataSource, mergeKey, version, data);

        // Get old version to know which one to update
    	var oldV = Optional.ofNullable(versionInformations.get(vd.getVersionKey()));
    	// Add new state to idx
        idx.put(oldV, vd, versionInformationsIdxEnvState);

        versionInformations.put(vd.getVersionKey(), vd);
    }

    public void updateVersion(VersionedData vd) {
        // Get old version to know which one to update
    	var oldV = Optional.ofNullable(versionInformations.get(vd.getVersionKey()));
    	// Add new state to idx
        idx.put(oldV, vd, versionInformationsIdxEnvState);

        versionInformations.put(vd.getVersionKey(), vd);
    }

	@Override
	public Optional<List<VersionedData>> getLatestVersion(String idxName, String attribute1, String attribute2) {

		 Optional<Integer> key = idx.getKey(attribute1, attribute2);

		 if(key.isEmpty()) {
			 logger.debug("Error while creating key from inputdata");
			 return Optional.empty();
		 } else {
			var vds = Optional.ofNullable(versionInformationsIdxEnvState.get(key.get()))
			.map(Map::values);


			if(vds.isEmpty()) {
				return Optional.empty();
			} else {
				return Optional.of(new ArrayList(vds.get()));
			}

		 }
	}

    @Override
    public void persist(String path) {
        writeToFile(path);
    }

    @Override
    public void persist(OutputStream stream) {
        writeToStream(stream);
    }

    @Override
    public void load(String path) {
        versionInformations = readFromFile(path);
    }

    @Override
    public void load(InputStream stream) {
        versionInformations = readFromStream(stream);
    }

    @Override
    public List<VersionedData> getAllVersionedResources() {
        return new ArrayList<>(versionInformations.values());
    }

    // === Private helper methods ===

    private Map<Integer, VersionedData> readFromFile(String path) {
        try (BufferedReader reader = new BufferedReader(new FileReader(path))) {
            Type listType = new TypeToken<List<VersionedData>>() {}.getType();
            List<VersionedData> dataList = gson.fromJson(reader, listType);
            return toMap(dataList != null ? dataList : new ArrayList<>());
        } catch (IOException e) {
            logger.debug("Unable to open file {}. Using empty version collection", path, e);
            return new HashMap<>();
        }
    }

    private Map<Integer, VersionedData> readFromStream(InputStream stream) {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            Type listType = new TypeToken<List<VersionedData>>() {}.getType();
            List<VersionedData> dataList = gson.fromJson(reader, listType);
            return toMap(dataList != null ? dataList : new ArrayList<>());
        } catch (IOException e) {
            logger.debug("Unable to read from stream. Using empty version collection", e);
            return new HashMap<>();
        }
    }

    private Map<Integer, VersionedData> toMap(List<VersionedData> dataList) {
        Map<Integer, VersionedData> resultMap = new ConcurrentHashMap<>();
        for (VersionedData data : dataList) {
            int key = data.getVersionKey();
            VersionedData existing = resultMap.get(key);
            if (existing == null || data.getVersion() >= existing.getVersion()) {
                resultMap.put(key, data);
            } else {
                logger.debug("Existing version is newer: existing={}, incoming={}", existing, data);
            }
        }
        return resultMap;
    }

    private void writeToFile(String path) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(path))) {
            gson.toJson(getAllVersionedResources(), writer);
        } catch (IOException e) {
            logger.error("Failed to write to file: {}", path, e);
        }
    }

    private void writeToStream(OutputStream stream) {
        try (Writer writer = new OutputStreamWriter(stream, StandardCharsets.UTF_8)) {
            gson.toJson(getAllVersionedResources(), writer);
            writer.flush();
        } catch (IOException e) {
            logger.error("Failed to write to output stream", e);
        }
    }
}
